src/base/utils/tar.cpp

```diff
@@ -0,0 +1,248 @@
+        // If the 1st byte is null then we can't extract a name. So the whole header becomes invalid to us.
+        if ((state == State::Valid) && header.startsWith('\0'))
+            return State::Invalid;
+        else if (state != State::Valid)
+            return state;
+
+        fileSize = octalToDec(Utils::ByteArray::midView(header, TAR_FILESIZE_OFFSET, TAR_FILESIZE_SIZE));
+        entryType = static_cast<Utils::Tar::StreamReader::EntryType>(header.at(TAR_FILETYPE_OFFSET));
+
+        const QByteArray name = Utils::ByteArray::midView(header, TAR_FILENAME_OFFSET, TAR_FILENAME_SIZE);
+        const int nameLen = strlen(name);
+        if (nameLen == 0)
+            return State::Invalid;
+
+        const QByteArray namePrefix = Utils::ByteArray::midView(header, TAR_FILENAME_PREFIX_OFFSET, TAR_FILENAME_PREFIX_SIZE);
+        const int namePrefixLen = strlen(namePrefix);
+
+        const QString finalName = (namePrefixLen > 0) ?
+                                      QString::fromUtf8(namePrefix.constData(), namePrefixLen)
+                                      + '/' + QString::fromUtf8(name.constData(), nameLen)
+                                    : QString::fromUtf8(name.constData(), nameLen);
+
+        fileName = Utils::Fs::toValidFileSystemName(finalName, true);
+
+        return State::Valid;
+    }
+}
+
+QHash<QString, QByteArray> Utils::Tar::untar(const QByteArray &data, QString &errorString)
+{
+    QHash<QString, QByteArray> files;
+    StreamReader reader(data);
+    while (reader.readNext()) {
+        if (!(reader.entryType() == StreamReader::EntryType::FileAscii
+              || reader.entryType() == StreamReader::EntryType::FileNull))
+            continue;
+
+        files.insert(reader.fileName(), reader.fileData());
+    }
+
+    errorString = reader.errorString();
+    return files;
+}
+
+Utils::Tar::StreamReader::StreamReader(const QByteArray &data)
+    : m_data(data)
+{
+    if (m_data.size() < TAR_HEADER_SIZE)
+        m_errorString = QCoreApplication::translate
+                        ("Utils::Tar",
+                         "The tar file is only %1 bytes long.",
+                         "'tar' should be left untranslated. It is a filetype.")
+                        .arg(m_data.size());
+}
+
+bool Utils::Tar::StreamReader::readNext()
+{
+    if (atEnd() || hasError())
+        return false;
+
+    int dataSize = m_data.size();
+    if (m_pos >= dataSize)
+        return false;
+
+    if (dataSize - m_pos < TAR_HEADER_SIZE) {
+        m_errorString = QCoreApplication::translate
+                        ("Utils::Tar",
+                         "The tar file has ended prematurely",
+                         "'tar' should be left untranslated. It is a filetype.");
+        return false;
+    }
+
+    const int padding = (m_fileSize > 0) ? (TAR_HEADER_SIZE - (m_fileSize % TAR_HEADER_SIZE))
+                                       : 0;
+
+    if (m_hasParsingStarted)
```

Comment:
I know.
It was either this `bool` or an `int` to hold to start position of the filedata. I chose bool in an effort to minimize class size in memory. But now I see that my last bool simply introduces padding at the end of the class. So instead of padding I'll use the `int` which should fit nicely there.

Reply:
> I chose bool in an effort to minimize class size in memory. 

It's invalid reason to make logic inconsistent.